### ISSUE TITLE
error using OAuth2 wih getmail-gmail-xoauth-tokens

### ISSUE THREAD

**Issue by al0git** [getmail6/getmail6#124]
Hello,

I'm trying to use OAuth2 with a Microsoft account and get errors. I'm following the documentation here: https://www3.isi.edu/~johnh/OTHER/LINUX/OAUTH2/index.html and example 13 in getmailrc-examples and get the following error during the init phase: urllib.error.HTTPError: HTTP Error 401: Unauthorized

I'm using onedrive with a 2FA authentication without problems. 

Any ideas ? 

Thanks.

I am on a Debian testing system. Here's the log:

% python3 /usr/bin/getmail-gmail-xoauth-tokens --init   ./microsoft-provider-thunderbird.json 
Visit this url to obtain a verification code:
    https://login.microsoftonline.com/common/oauth2/v2.0/authorize?access_type=offline&client_id=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&prompt=consent&redirect_uri=http%3A%2F%2Flocalhost%3A8083%2F&response_type=code&scope=offline_access%20https%3A%2F%2Foutlook.office.com%2FIMAP.AccessAsUser.All%20https%3A%2F%2Foutlook.office.com%2FPOP.AccessAsUser.All%20https%3A%2F%2Foutlook.office.com%2FSMTP.Send

127.0.0.1 - - [27/Jul/2022 21:58:21] "GET /?code=XXXXXXXXXXXXXXXXXX.....XXXXXXXX&session_state=e92ad1e2-5858-413b-8030-06c5d28c7f7c HTTP/1.1" 200 -
Traceback (most recent call last):
  File "/usr/bin/getmail-gmail-xoauth-tokens", line 148, in <module>
    auth.init_tokens(oauthd.oauth_code,args.port)
  File "/usr/bin/getmail-gmail-xoauth-tokens", line 113, in init_tokens
    d = self.get_response(self.data['token_uri'], params)
  File "/usr/bin/getmail-gmail-xoauth-tokens", line 93, in get_response
    response = urllibrequest.urlopen(url, encoded).read()
  File "/usr/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)
  File "/usr/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 401: Unauthorized

The json file contains:
{
    "user": "XXXX@YYYY.com",
    "scope": "offline_access https://outlook.office.com/IMAP.AccessAsUser.All https://outlook.office.com/POP.AccessAsUser.All https://outlook.office.com/SMTP.Send",
    "client_id": "XXXXXXXXXXXXXXXXXX",
    "client_secret": "code=xxxxxxxx",
    "tenant": "common",
    "auth_uri": "https://login.microsoftonline.com/common/oauth2/v2.0/authorize",
    "token_uri": "https://login.microsoftonline.com/common/oauth2/v2.0/token",
 	"redirect_uri": "http://localhost:8083"
}

**Comment by ykasap** [comment]
It seems that your "client_secret" in the JSON file is wrong.  It should contain a string provided by Azure AD when you created a new client secret after you registered your instance of getmail6 as an application.  It should not start with "code=".

**Comment by ChrisTG742** [comment]
The docs are not yet updated as it seems: One no-longer needs to copy&past the code, but forward the port 8083 to the local machine, so that the redirect that is created when calling the URL reaches the host running getmail again. The needed modifications are then done automatically by the script. The old copy&paste method is no longer needed and also does not work any more. Maybe that's the problem you are facing?

**Comment by al0git** [comment]
Hello,
Sorry for not responding sooner, had real holidays.
I tried modifying the client_secret, but it didn't work either.
As I was really in a hurry (no more mails), I switched to using the mutt oauth script which worked.
Thanks.

**Comment by ChrisTG742** [comment]
Try to remove the "prompt=consent" from the URL. It seems, that this leads to the additional redirect that makes the script crash:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?access_type=offline&client_id=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&redirect_uri=http%3A%2F%2Flocalhost%3A8083%2F&response_type=code&scope=offline_access%20https%3A%2F%2Foutlook.office.com%2FIMAP.AccessAsUser.All%20https%3A%2F%2Foutlook.office.com%2FPOP.AccessAsUser.All%20https%3A%2F%2Foutlook.office.com%2FSMTP.Send

**Comment by dequb** [comment]
Hi, I'm having the same issues with Microsoft. My json file looks like this:

```
{
 "scope": "https://outlook.office365.com/IMAP.AccessAsUser.All https://outlook.office365.com/POP.AccessAsUser.All https://outlook.office365.com/SMTP.Send offline_access",
  "user": "(my username)",
  "client_id": "(client id)",
  "client_secret": "(secret - the string under value and not the secret ID)",
  "token_uri": "https://login.microsoftonline.com/(tenant)/oauth2/v2.0/token",
  "auth_uri": "https://login.microsoftonline.com/(tenant)/oauth2/v2.0/authorize",
  "redirect_uri": "http://localhost:8083"
}
```

I double checked the client_id, secret, tenant as well as the permission in the Azure panel, but I'm always getting this in the console:

```
# getmail-gmail-xoauth-tokens -i provider.json 
Visit this url to obtain a verification code:
    https://login.microsoftonline.com/(long url)

127.0.0.1 - - [19/Sep/2022 12:55:06] "GET /?code=(very long string) HTTP/1.1" 200 -
Traceback (most recent call last):
  File "/usr/bin/getmail-gmail-xoauth-tokens", line 148, in <module>
    auth.init_tokens(oauthd.oauth_code,args.port)
  File "/usr/bin/getmail-gmail-xoauth-tokens", line 113, in init_tokens
    d = self.get_response(self.data['token_uri'], params)
  File "/usr/bin/getmail-gmail-xoauth-tokens", line 93, in get_response
    response = urllibrequest.urlopen(url, encoded).read()
  File "/usr/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)
  File "/usr/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 401: Unauthorized
```

The browser says "Your json file is updated" but there is no update.

Any help? Thanks!

**Comment by TIS-Stefan** [comment]
Hello

Same error here. Even with and without prompt=consent. 
I am trying since 4 hours now without success.
Neither with Python 2.7, nor with 3.8.

Stefan

**Comment by dequb** [comment]
Quick update: I got it working removing the secret from the "Client secrets" in Azure and setting client_secret to "".

**Comment by TIS-Stefan** [comment]
Well, indeed, it works. But it somehow does not sound for a good plan. 

Stefan

**Comment by TIS-Stefan** [comment]
Just an annotation: 
Without the "client_secret", I do not now, when the token expires. 
I suppose, getmail-gmail-xoauth-tokens makes something wrong with the request at the client secret. Unfortunately I have no time to look into that now. 

Stefan

**Comment by ChrisTG742** [comment]
Actually, without "consent=prompt" it worked for me here. The token were written to the JSON file and afterwards the auth worked flawlessly.
So I rather think this depends on how you set up the app registration in Azure. Maybe some of your settings are different to mine?

**Comment by TIS-Stefan** [comment]
I commented the line

` # params['prompt'] = 'consent'`

from "def get_response".
Added a secret. But I always get the 401 not authorized. 

I also checked the params in "init_tokens" and they look ok to me. 

Without the client_secret, the token expires after one hour and is not refreshed.  getmail-gmail-xoauth-tokens returns error 1.

The settings in Azure are according to the documentation at https://github.com/getmail6/getmail6/blob/master/docs/getmailrc-examples line 351. Which does not mean I forgot something.

I think, you are right, there is something different in the registration in Azure.

BTW: 
line 105
`
json.dump(self.data, f, indent=4)
`
makes the json better readable.
